File: app/src/main/java/d/d/meshenger/StartActivity.java
package d.d.meshenger;

import android.app.Dialog;
import android.app.Service;
import android.content.ComponentName;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.graphics.Typeface;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatDelegate;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import org.libsodium.jni.Sodium;
import org.libsodium.jni.SodiumConstants;

import java.util.ArrayList;

import org.libsodium.jni.NaCl;


public class StartActivity extends MeshengerActivity implements ServiceConnection {
    private MainService.MainBinder binder;
    private int startState = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_splash);

        // load libsodium for JNI access
        NaCl.sodium();

        Typeface type = Typeface.createFromAsset(getAssets(), "rounds_black.otf");
        ((TextView) findViewById(R.id.splashText)).setTypeface(type);

        // start MainService and call back via onServiceConnected()
        startService(new Intent(this, MainService.class));
    }

    private void continueInit() {
        this.startState += 1;

        switch (this.startState) {
            case 1:
                log("init 1: load database");
                // open without password
                this.binder.loadDatabase();
                continueInit();
                break;
            case 2:
                log("init 2: check database");
                if (this.binder.getDatabase() == null) {
                    // database is probably encrypted
                    showDatabasePasswordDialog();
                } else {
                    continueInit();
                }
                break;
            case 3:
                log("init 3: check username");
                if (this.binder.getSettings().getUsername().isEmpty()) {
                    // set username
                    showMissingUsernameDialog();
                } else {
                    continueInit();
                }
                break;
            case 4:
                log("init 4: check key pair");
                if (this.binder.getSettings().getPublicKey().isEmpty()) {
                    // generate key pair
                    setKeyPair();
                }
                continueInit();
                break;
            case 5:
                log("init 5: check addresses");
                if (this.binder.isFirstStart()) {
                    showMissingAddressDialog();
                } else {
                    continueInit();
                }
                break;
            case 6:
               log("init 6: start contact list");
                // set night mode
                boolean nightMode = this.binder.getSettings().getNightMode();
                AppCompatDelegate.setDefaultNightMode(
                        nightMode ? AppCompatDelegate.MODE_NIGHT_YES : AppCompatDelegate.MODE_NIGHT_NO
                );

                // all done - show contact list
                startActivity(new Intent(this, ContactListActivity.class));
                finish();
                break;
        }
    }

    @Override
    public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
        this.binder = (MainService.MainBinder) iBinder;

        if (this.binder.isFirstStart()) {
            // show delayed splash page
            (new Handler()).postDelayed(() -> { continueInit(); }, 1000);
        } else {
            // show contact list as fast as possible
            continueInit();
        }
    }

    @Override
    public void onServiceDisconnected(ComponentName componentName) {
        this.binder = null;
    }

    @Override
    protected void onResume() {
        super.onResume();
        bindService(new Intent(this, MainService.class), this, Service.BIND_AUTO_CREATE);
    }

    @Override
    protected void onPause() {
        super.onPause();
        unbindService(this);
    }

    private void setKeyPair() {
        // create secret/public key pair
        byte[] publicKey = new byte[SodiumConstants.PUBLICKEY_BYTES];
        byte[] secretKey = new byte[SodiumConstants.SECRETKEY_BYTES];
        Sodium.crypto_box_keypair(publicKey, secretKey);

        Settings settings = this.binder.getSettings();
        settings.setPublicKey(Utils.byteArrayToHexString(publicKey).toUpperCase());
        settings.setSecretKey(Utils.byteArrayToHexString(secretKey).toUpperCase());

        try {
            this.binder.saveDatabase();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void showMissingAddressDialog() {
        ArrayList<String> addresses = Utils.getMacAddresses();
        if (addresses.isEmpty()) {
            Toast.makeText(this, "No contact address found! Please configure.", Toast.LENGTH_LONG).show();
            continueInit();
        } else {
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle("Setup Address");
            builder.setMessage("No address for your QR-Code configured. Set to: " + Utils.join(addresses));

            builder.setPositiveButton(R.string.ok, (DialogInterface dialog, int id) -> {
                for (String address : addresses) {
                    this.binder.getSettings().addAddress(address);
                }

                try {
                    this.binder.saveDatabase();
                } catch (Exception e) {
                    e.printStackTrace();
                }

                dialog.cancel();
                continueInit();
            });

            builder.setNegativeButton(R.string.skip, (DialogInterface dialog, int id) -> {
                dialog.cancel();
                // continue with out address configuration
                continueInit();
            });

            builder.show();
        }
    }

    // initial dialog to set the username
    private void showMissingUsernameDialog() {
        TextView tw = new TextView(this);
        tw.setText(R.string.name_prompt);
        //tw.setTextColor(Color.BLACK);
        tw.setTextSize(20);
        tw.setGravity(Gravity.CENTER_HORIZONTAL);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.addView(tw);

        EditText et = new EditText(this);
        et.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        et.setSingleLine(true);

        layout.addView(et);
        layout.setPadding(40, 80, 40, 40);
        //layout.setGravity(Gravity.CENTER_HORIZONTAL);

        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle(R.string.hello);
        builder.setView(layout);
        builder.setNegativeButton(R.string.cancel, (dialogInterface, i) -> {
            this.binder.shutdown();
            finish();
        });

        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);

        builder.setPositiveButton(R.string.next, (dialogInterface, i) -> {
            // we will override this handler
        });

        AlertDialog dialog = builder.create();
        dialog.setOnShowListener((newDialog) -> {
            Button okButton = ((AlertDialog) newDialog).getButton(AlertDialog.BUTTON_POSITIVE);
            et.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    // nothing to do
                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                    // nothing to do
                }

                @Override
                public void afterTextChanged(Editable editable) {
                    okButton.setClickable(editable.length() > 0);
                    okButton.setAlpha(editable.length() > 0 ? 1.0f : 0.5f);
                }
            });

            okButton.setClickable(false);
            okButton.setAlpha(0.5f);

            if (et.requestFocus()) {
                imm.showSoftInput(et, InputMethodManager.SHOW_IMPLICIT);
                //imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
            }
        });

        dialog.show();

        // override handler (to be able to dismiss the dialog manually)
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener((View v) -> {
            imm.toggleSoftInput(0, InputMethodManager.HIDE_IMPLICIT_ONLY);

            String username = et.getText().toString();
            this.binder.getSettings().setUsername(username);

            try {
                this.binder.saveDatabase();
            } catch (Exception e) {
                e.printStackTrace();
            }

            // close dialog
            dialog.dismiss();
            //dialog.cancel(); // needed?
            continueInit();
        });
    }

    // ask for database password
    private void showDatabasePasswordDialog() {
        Dialog dialog = new Dialog(this);
        dialog.setContentView(R.layout.database_password_dialog);
        //dialog.setTitle("Title...");

        EditText passwordEditText = dialog.findViewById(R.id.PasswordEditText);
        Button exitButton = dialog.findViewById(R.id.ExitButton);
        Button okButton = dialog.findViewById(R.id.OkButton);

        okButton.setOnClickListener((View v) -> {
            String password = passwordEditText.getText().toString();
            this.binder.setDatabasePassword(password);
            this.binder.loadDatabase();

            if (this.binder.getDatabase() == null) {
                Toast.makeText(this, R.string.wrong_password, Toast.LENGTH_SHORT).show();
            } else {
                // close dialog
                dialog.dismiss();
                continueInit();
            }
        });

        exitButton.setOnClickListener((View v) -> {
            // shutdown app
            dialog.dismiss();
            this.binder.shutdown();
            finish();
        });

        dialog.show();
    }

    private void log(String s) {
        Log.d(StartActivity.class.getSimpleName(), s);
    }
}

File: app/src/main/java/d/d/meshenger/StartActivity.kt
package d.d.meshenger

import android.app.Dialog
import android.content.ComponentName
import android.content.DialogInterface
import android.content.Intent
import android.content.ServiceConnection
import android.graphics.Typeface
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.text.Editable
import android.text.TextWatcher
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import d.d.meshenger.MainService
import d.d.meshenger.MainService.MainBinder
import org.libsodium.jni.NaCl
import org.libsodium.jni.Sodium
import java.util.*

/*
 * Show splash screen, name setup dialog, database password dialog and
 * start background service before starting the MainActivity.
 */
class StartActivity : MeshengerActivity(), ServiceConnection {
    private var binder: MainBinder? = null
    private var dialog : Dialog? = null
    private var startState = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)
        // load libsodium for JNI access
        sodium = NaCl.sodium()

        val type = Typeface.createFromAsset(assets, "rounds_black.otf")
        findViewById<TextView>(R.id.splashText).setTypeface(type)

        // start MainService and call back via onServiceConnected()
        startService(Intent(this, MainService::class.java))
    }

    private fun continueInit() {
        startState += 1
        when (startState) {
            1 -> {
                Log.d(this, "init 1: load database")
                // open without password
                binder!!.getService().loadDatabase()
                continueInit()
            }
            2 -> {
                Log.d(this, "init 2: check database")
                if (binder!!.getDatabase() == null) {
                    // database is probably encrypted
                    showDatabasePasswordDialog()
                } else {
                    continueInit()
                }
            }
            3 -> {
                Log.d(this, "init 3: check username")
                if (binder!!.getSettings().username.isEmpty()) {
                    // set username
                    showMissingUsernameDialog()
                } else {
                    continueInit()
                }
            }
            4 -> {
                Log.d(this, "init 4: check key pair")
                if (binder!!.getSettings().publicKey.isEmpty()) {
                    // generate key pair
                    initKeyPair()
                }
                continueInit()
            }
            5 -> {
                Log.d(this, "init 5: check addresses")
                if (binder!!.getService().first_start) {
                    showMissingAddressDialog()
                } else {
                    continueInit()
                }
            }
            6 -> {
                Log.d(this, "init 6: start contact list")
                // set night mode
                val nightMode = binder!!.getSettings().nightMode
                AppCompatDelegate.setDefaultNightMode(
                    if (nightMode) AppCompatDelegate.MODE_NIGHT_YES else AppCompatDelegate.MODE_NIGHT_NO
                )
                // all done - show contact list
                startActivity(Intent(this, MainActivity::class.java))
                finish()
            }
        }
    }

    override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
        binder = iBinder as MainBinder
        Log.d(this, "onServiceConnected")
        if (startState == 0) {
            if (binder!!.getService().first_start) {
                // show delayed splash page
                Handler().postDelayed({ continueInit() }, 1000)
            } else {
                // show contact list as fast as possible
                continueInit()
            }
        }
    }

    override fun onServiceDisconnected(componentName: ComponentName) {
        binder = null
    }

    override fun onResume() {
        super.onResume()
        bindService(Intent(this, MainService::class.java), this, BIND_AUTO_CREATE)
    }

    override fun onPause() {
        super.onPause()
        unbindService(this)
    }

    private fun initKeyPair() {
        // create secret/public key pair
        val publicKey = ByteArray(Sodium.crypto_sign_publickeybytes())
        val secretKey = ByteArray(Sodium.crypto_sign_secretkeybytes())
        Sodium.crypto_sign_keypair(publicKey, secretKey)
        val settings = binder!!.getSettings()
        settings.publicKey = publicKey
        settings.secretKey = secretKey
        try {
            binder!!.saveDatabase()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun getDefaultAddress(): AddressEntry? {
        val addresses = Utils.collectAddresses()

        // preferable, since we can device fe80:: and other addresses from it
        val macAddress = addresses.firstOrNull { it.device.startsWith("wlan") && Utils.isMACAddress(it.address) }
        if (macAddress != null) {
            return macAddress
        }

        val fe80Address = addresses.firstOrNull { it.device.startsWith("wlan") && it.address.startsWith("fe80::") }
        if (fe80Address != null) {
            return fe80Address
        }

        return null
    }

    private fun showMissingAddressDialog() {
        val defaultAddress = getDefaultAddress()
        if (defaultAddress == null) {
            val builder = AlertDialog.Builder(this)
            builder.setTitle("Setup Address")
            builder.setMessage("No address of your WiFi card found. Enable WiFi now (no Internet needed) or skip to configure later.")
            builder.setPositiveButton(R.string.ok) { dialog: DialogInterface, _: Int ->
                showMissingAddressDialog()
                dialog.cancel()
            }
            builder.setNegativeButton(R.string.skip) { dialog: DialogInterface, _: Int ->
                dialog.cancel()
                // continue with out address configuration
                continueInit()
            }
            builder.show()
        } else {
            binder!!.getSettings().addresses = mutableListOf(defaultAddress.address)
            try {
                binder!!.saveDatabase()
            } catch (e: Exception) {
                e.printStackTrace()
            }
            continueInit()
        }
    }

    // initial dialog to set the username
    private fun showMissingUsernameDialog() {
        val tw = TextView(this)
        tw.setText(R.string.name_prompt)
        //tw.setTextColor(Color.BLACK);
        tw.textSize = 20f
        tw.gravity = Gravity.CENTER_HORIZONTAL
        val layout = LinearLayout(this)
        layout.orientation = LinearLayout.VERTICAL
        layout.addView(tw)
        val et = EditText(this)
        et.layoutParams = LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        )
        et.isSingleLine = true
        layout.addView(et)
        layout.setPadding(40, 80, 40, 40)
        //layout.setGravity(Gravity.CENTER_HORIZONTAL);
        val builder = AlertDialog.Builder(this)
        builder.setTitle(R.string.hello)
        builder.setView(layout)
        builder.setNegativeButton(R.string.skip) { _: DialogInterface?, _: Int ->
            binder!!.shutdown()
            finish()
        }
        val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
        builder.setPositiveButton(R.string.next) { _: DialogInterface?, _: Int -> }
        val dialog = builder.create()
        dialog.setOnShowListener { newDialog: DialogInterface ->
            val okButton = (newDialog as AlertDialog).getButton(
                AlertDialog.BUTTON_POSITIVE
            )
            et.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    charSequence: CharSequence,
                    i: Int,
                    i1: Int,
                    i2: Int,
                ) {
                    // nothing to do
                }

                override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                    // nothing to do
                }

                override fun afterTextChanged(editable: Editable) {
                    okButton.isClickable = editable.isNotEmpty()
                    okButton.alpha = if (editable.isNotEmpty()) 1.0f else 0.5f
                }
            })
            okButton.isClickable = false
            okButton.alpha = 0.5f
            if (et.requestFocus()) {
                imm.showSoftInput(et, InputMethodManager.SHOW_IMPLICIT)
                //imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
            }
        }
        dialog.show()
        // override handler (to be able to dismiss the dialog manually)
        dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
            imm.toggleSoftInput(0, InputMethodManager.HIDE_IMPLICIT_ONLY)
            val username = et.text.toString()
            if (Utils.isValidName(username)) {
                binder!!.getSettings().username = username
                try {
                    binder!!.saveDatabase()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                // close dialog
                dialog.dismiss()
                //dialog.cancel(); // needed?
                continueInit()
            } else {
                Toast.makeText(this, R.string.invalid_name, Toast.LENGTH_SHORT).show()
            }
        }
        dialog.getButton(AlertDialog.BUTTON_NEGATIVE).setOnClickListener {
            imm.toggleSoftInput(0, InputMethodManager.HIDE_IMPLICIT_ONLY)
            val username = generateRandomUserName()
            if (Utils.isValidName(username)) {
                binder!!.getSettings().username = username
                try {
                    binder!!.saveDatabase()
                } catch (e: Exception) {
                    e.printStackTrace()
                }
                // close dialog
                dialog.dismiss()
                //dialog.cancel(); // needed?
                continueInit()
            } else {
                Toast.makeText(this, R.string.invalid_name, Toast.LENGTH_SHORT).show()
            }
        }
    }

    // ask for database password
    private fun showDatabasePasswordDialog() {
        val dialog = Dialog(this)
        dialog.setContentView(R.layout.dialog_database_password)
        val passwordEditText = dialog.findViewById<EditText>(R.id.change_password_edit_textview)
        val exitButton = dialog.findViewById<Button>(R.id.change_password_cancel_button)
        val okButton = dialog.findViewById<Button>(R.id.change_password_ok_button)
        okButton.setOnClickListener {
            val password = passwordEditText.text.toString()
            binder!!.getService().database_password = password
            binder!!.getService().loadDatabase()
            if (binder!!.getDatabase() == null) {
                Toast.makeText(this, R.string.wrong_password, Toast.LENGTH_SHORT).show()
            } else {
                // close dialog
                dialog.dismiss()
                continueInit()
            }
        }
        exitButton.setOnClickListener {
            // shutdown app
            dialog.dismiss()
            binder!!.shutdown()
            finish()
        }
        dialog.show()
    }

    companion object {
        private var sodium: Sodium? = null
    }

    private fun generateRandomUserName(): String {
        val user = getString(R.string.user)
        val id = UUID.randomUUID().toString().substring(0..6)
        return "$user-$id"
    }
}